datafusion/src/execution/memory_management/mod.rs

```diff
@@ -0,0 +1,405 @@
+                    self.partition_id,
+                    released,
+                    consumer.str_repr()
+                );
+                got += memory_manager
+                    .acquire_exec_pool_memory(required - got, consumer_id)
+                    .await;
+            }
+        }
+
+        if got < required {
+            return Err(OutOfMemory(format!(
+                "Unable to acquire {} bytes of memory, got {}",
+                required, got
+            )));
+        }
+
+        debug!("{} acquired {}", consumer_id, got);
+        Ok(got)
+    }
+
+    /// log current memory usage for all consumers in this partition
+    pub async fn show_memory_usage(&self) -> Result<()> {
+        info!("Memory usage for partition {}", self.partition_id);
+        let consumers = self.consumers.lock().await;
+        let mut used = 0;
+        for (_, c) in consumers.iter() {
+            let cur_used = c.get_used();
+            used += cur_used;
+            if cur_used > 0 {
+                info!(
+                    "Consumer {} acquired {}",
+                    c.str_repr(),
+                    human_readable_size(cur_used as usize)
+                )
+            }
+        }
+        let no_consumer_size = self
+            .memory_manager
+            .upgrade()
+            .ok_or_else(|| {
+                DataFusionError::Execution("Failed to get MemoryManager".to_string())
+            })?
+            .exec_memory_used_for_partition(self.partition_id)
+            - (used as usize);
+        info!(
+            "{} bytes of memory were used for partition {} without specific consumer",
+            human_readable_size(no_consumer_size),
+            self.partition_id
+        );
+        Ok(())
+    }
+}
+
+#[derive(Clone, Debug, Hash, Eq, PartialEq)]
+/// Id that uniquely identifies a Memory Consumer
+pub struct MemoryConsumerId {
+    /// partition the consumer belongs to
+    pub partition_id: usize,
+    /// unique id
+    pub id: usize,
+}
+
+impl MemoryConsumerId {
+    /// Auto incremented new Id
+    pub fn new(partition_id: usize) -> Self {
+        let id = next_id();
+        Self { partition_id, id }
+    }
+}
+
+impl Display for MemoryConsumerId {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{}:{}", self.partition_id, self.id)
+    }
+}
+
+#[async_trait]
+/// A memory consumer that supports spilling.
+pub trait MemoryConsumer: Send + Sync + Debug {
+    /// Display name of the consumer
+    fn name(&self) -> String;
+
+    /// Unique id of the consumer
+    fn id(&self) -> &MemoryConsumerId;
+
+    /// Ptr to MemoryManager
+    fn memory_manager(&self) -> Arc<MemoryManager>;
+
+    /// partition that the consumer belongs to
+    fn partition_id(&self) -> usize {
+        self.id().partition_id
+    }
+
+    /// Try allocate `required` bytes as needed
+    async fn allocate(&self, required: usize) -> Result<()> {
```

Comment:
Have you considered some sort of RAII mechanism here (to help ensure allocations match up with deallocations)?

Something like

```suggestion
    async fn allocate(&self, required: usize) -> Result<ResourceReservation> {
```

and 
```rust
struct ResourceReservation {...}


impl Drop for ResourceReservation {
  fn drop(&mut self) -> { 
    // release the reservation
  }
}
```

Reply:
I haven't thought about it before, great idea👍